### PR TITLE
feat: allow no lower bound gap fill implementation

### LINKED ISSUE #7078: Make `GapFill` accept implied bounds for InfluxQL compatability
InfluxQL will allow no time range in the WHERE clause when doing a gap-filling query, e.g:
```sql
select k0, k1, max(f0) from m group by k0, k1, time(1m) fill(previous);
name: m
tags: k0=foo, k1=bar
time                 k0  k1  max
----                 --  --  ---
2023-02-27T21:59:00Z foo bar 100
2023-02-27T22:00:00Z foo bar 100
2023-02-27T22:01:00Z foo bar 100
2023-02-27T22:02:00Z foo bar 100
2023-02-27T22:03:00Z foo bar 100
2023-02-27T22:04:00Z foo bar 100
```
In this case the implied bounds for gap-filling are:
- start with the window of the first timestamp in the series
- end with the window of `now()`

Currently the GapFill operator will panic if this is attempted:
https://github.com/influxdata/influxdb_iox/blob/22fe629f549e720c733459cfa2dbdd10b89a11a0/iox_query/src/exec/gapfill.rs#L63-L68

Likewise there will need to be changes to the implementation.

### PR DESCRIPTION
**Pull request by wolffcm**
Closes #7078.

This updates the gap-filling operator to be able to function without a lower bound in the time range supplied by the caller. This is needed for the InfluxQL front end, which will just use the first timestamp in the series as the lower bound if one is not present in the WHERE clause of the query.

Note: the IOx InfluxQL planner is not able to do gap filling yet, but in the near future it can be extended to add a `GapFill` node to the plan to achieve this.

I opted not to make the upper bound optional for gap filling:
- InfluxQL allows the upper bound to be implicit in gap filling queries, but in this case the IOx InfluxQL planner can provide a call to `now()` [here](https://github.com/influxdata/influxdb_iox/blob/525c48de2c758ac406e1b5deee68982fa9ba95ac/iox_query/src/exec/gapfill.rs#L52) which will be evaluated by the DataFusion planner.
- Making the upper bound required simplifies the operator implementation.

The changes in this PR at a high-level:
- Update the logical node `GapFill` to allow `Unbounded` as the start of the time range, and associated tests
- Update `gapfill::params::GapFillParams` to have an optional first timestamp.
- Update the operator implementation in `algo.rs` to have logic to choose the first timestamp for a series depending on the parameters.

### CODE REVIEW

File: iox_query/src/exec/gapfill.rs [hunk 1]
```diff
@@ -909,6 +961,59 @@ mod test {
...
+                    agg_cols: vec![vec![
+                        Some(10), // group a
+                        Some(90), // gropu b
+                        Some(91),
+                        Some(92),
+                        Some(93),
+                    ]],
+                    input_batch_size,
+                };
+                let params = get_params_ms(&records, 25, Some(975), 1_025);
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill.rs:987]: This new test isn't really relevant to the work of this PR, but I realized I was missing a test case where there is a series that has only null timestamps, which seemed like a corner case worth testing.

File: iox_query/src/exec/gapfill.rs [hunk 2]
```diff
@@ -48,24 +48,29 @@ pub(crate) struct GapFillParams {
     /// The origin argument from the call to DATE_BIN_GAPFILL
     pub origin: Expr,
     /// The time range of the time column inferred from predicates
-    /// in the overall query
+    /// in the overall query. The lower bound may be [`Bound::Unbounded`]
+    /// which implies that gap-fililng should just start from the
```
Comment (stuartcarnie) [inline iox_query/src/exec/gapfill.rs:52]:
```suggestion
    /// which implies that gap-filling should just start from the
```

File: iox_query/src/exec/gapfill.rs [hunk 3]
```diff
@@ -48,24 +48,29 @@ pub(crate) struct GapFillParams {
...
-            bound_extract(&self.time_range.start)
-                .unwrap_or_else(|| panic!("lower time bound is required"))
-                .clone(),
+        ];
+        if let Some(start) = bound_extract(&self.time_range.start) {
+            exprs.push(start.clone());
+        }
+        exprs.push(
             bound_extract(&self.time_range.end)
                 .unwrap_or_else(|| panic!("upper time bound is required"))
```
Comment (stuartcarnie) [inline iox_query/src/exec/gapfill.rs:70]: Presumably this is an invariant that is guaranteed elsewhere, and so a `panic` here is an exceptional case that represents a programming error?

File: iox_query/src/exec/gapfill.rs [hunk 4]
```diff
@@ -909,6 +961,59 @@ mod test {
...
+                    time_col: vec![
+                        Some(1_000),
+                        None, // group b
+                        None,
+                        None,
+                        None,
+                    ],
+                    agg_cols: vec![vec![
+                        Some(10), // group a
+                        Some(90), // gropu b
```
Comment (stuartcarnie) [inline iox_query/src/exec/gapfill.rs:980]:
```suggestion
                        Some(90), // group b
```

File: iox_query/src/exec/gapfill/params.rs [hunk 5]
```diff
@@ -211,21 +227,58 @@ mod tests {
...
+            stride: interval(1_000_000_000),
+            time_column: Column::new("time", 0),
+            origin: timestamp(0),
+            time_range: Range {
+                start: Bound::Unbounded,
+                end: Bound::Excluded(timestamp(20_000_000_000)),
+            },
+        };
+
+        let actual = GapFillParams::try_new(schema().into(), &exec_params)?;
```
Comment (stuartcarnie) [inline iox_query/src/exec/gapfill/params.rs:263]: Not a requirement, just an alternative. 

```suggestion
        let actual = GapFillParams::try_new(schema().into(), &exec_params).unwrap();
```

You can `unwrap()` the result, which avoids having to add `-> Result<()>` and the `Ok(())` at the end of the test. A panic will occur on the line of the `unwrap`, if that condition fails, which is helpful for debugging. This tip was suggested to me by Marco.

**Review by stuartcarnie — APPROVED** [review]
Looks great Chris!

File: iox_query/src/exec/gapfill.rs (hunk 3 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill.rs:70]: Yes, that's right. `expressions` is called from `UserDefinedLogicalNode` impl for `GapFill`, whose `try_new` constructor will produce an error if the upper bound is missing. So this really should never happen.

Incidentally it is `expressions` paired with `from_template` that allows the DF planner to do things like evaluate constant expressions, which is pretty neat. You can define a bound like `now() - interval '10 minutes'` and at some point during planning DF replaces it with an appropriate literal.

File: iox_query/src/exec/gapfill/params.rs (hunk 5 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/params.rs:263]: I like it, I'll start writing tests this way.
